### PR TITLE
trying to implement the split function

### PR DESCRIPTION
**Pull request by koreaneggroll**
Also there is printSplit which prints out the new array it creates

### CODE REVIEW

**Comment by midnqp** [comment]
You've quite a courage trying new things out! That's immensely great & contagious!!

**Comment by koreaneggroll** [comment]
is that a good thing lmao ?

**Comment by koreaneggroll** [comment]
nvm just now understood what u meant lmao

File: libavoidrepetition.h [hunk 1]
```diff
@@ -158,3 +158,49 @@ double min(double numbers[], int arrlen) {
...
+
+
+size_t str_split(char ***array, char *str, const char *del) {
+    char *token;
+    size_t i = 0;
+
+    token = strtok_r(str, del);
+  
+    while(token != NULL) {
+        *array = realloc(*array, sizeof(char *) * (i + 1));
```
Comment (midnqp) [inline libavoidrepetition.h:171]: Hey Arin @koreaneggroll, we are **libavoidrepetition**. We strive at reusing code, and just avoid repetition.

Don't you think, in C, we always need to `realloc()` whenever we just want to add/concat strings? What about we make a new function `add(char** destinationStr, char* addingStr)`? What do you say?

Comment (koreaneggroll) [reply libavoidrepetition.h:171]: I don't think I fully grasp what you mean by that, could you go more in detail ?

File: libavoidrepetition.h [hunk 2]
```diff
@@ -158,3 +158,49 @@ double min(double numbers[], int arrlen) {
...
+    
+    for(size_t i = 0; i < count; i++) {
+        printf("[%s], ", split[i]);
+    }
+    
+    free(split);
+}
+
+
+char ***split(char **array, char *str, const char *del){
```
Comment (midnqp) [inline libavoidrepetition.h:192]: Amazing & Clever.

Comment (koreaneggroll) [reply libavoidrepetition.h:192]: thnx man

File: libavoidrepetition.h (hunk 1 above)
Comment (midnqp) [reply libavoidrepetition.h:171]: Uhh... Think of like:
`add(the_main_string_variable_name_here, the_string_that_will_be_added_to_main_string);`

Simple enough?

Comment (koreaneggroll) [reply libavoidrepetition.h:171]: So like, take the string that was split and copy that value to another variable?

**Comment by midnqp** [comment]
@koreaneggroll Would you mind adding tests in `__test__.c`?

I think I have added tests up to what is implemented until now.

**Comment by koreaneggroll** [comment]
yeah sure.

**Comment by koreaneggroll** [comment]
U mean like adding more tests with new functions or just testing the functions that are already there ?

**Comment by midnqp** [comment]
> U mean like adding more tests with new functions or just testing the functions that are already there ?

Adding test functions:
  - which can take user-input from `stdin`
  - and just invoke the implemented function from `libavoidrepetition.h`
  - and just print the returned results.

**Comment by koreaneggroll** [comment]
understood, on it

File: libavoidrepetition.h (hunk 1 above)
Comment (midnqp) [reply libavoidrepetition.h:171]: For anything, which adds two strings. Consider thinking of this example:
```C
char* httpResponse = (char*)malloc(1024);
char* httpHeader = (char*)malloc(1024);
char* httpBody = (char*)malloc(1024);

/* --code that gets some http header-- */
str_add(httpResponse, httpHeader);
/* --code that gets some response body(HTML)-- */
str_add(httpResponse, httpBody);
```

Comment (koreaneggroll) [reply libavoidrepetition.h:171]: Isn't that just strcat ?

Comment (midnqp) [reply libavoidrepetition.h:171]: Sure, it is -- unless we do the `malloc_reusable_size()` & `strlen()` checks, and `realloc()` accordingly. 

Sure we could just do:
`static char* httpResponse[100000];`

But I got injuriously ridiculed by a coder from reddit for using `[100000]`. And by using `static`, you agree to disallow operations on the string for future. So, seems like `malloc()` and `realloc()` is the only best practice.

Comment (koreaneggroll) [reply libavoidrepetition.h:171]: So you want to make a function that adds 2 strings togheter, and then make a third string and reallocate it memory and if it isn't big enough you reallocate it ? makes sense

Comment (midnqp) [reply libavoidrepetition.h:171]: Yes. The plan is:
`realloc()` returns an entire new pointer alongwith the string up to which is concatenated. So, we could: 
```C
char* addingStrTemp = (char*)realloc(...);
```
 and use that. And finally: 
```C
    ...
    ...
    return addingStrTemp;
} //add(strDest, strToAdd)
```

Comment (koreaneggroll) [reply libavoidrepetition.h:171]: well you don't need a function for that, even writing it from scratch is pretty simple ```c
    char *str1 = malloc(sizeof(char) * 1024);
    char *str2 = malloc(sizeof(char) * 1024);

    strcpy(str1, "whatever");
    strcpy(str2, " whatever once again");

    char *str3 = malloc(sizeof(str1) + sizeof(str2));

    strcat(str1, str2);
    strcpy(str3, str1);```

Comment (koreaneggroll) [reply libavoidrepetition.h:171]: don't know why that didn't put the code corectly inside but meh

Comment (midnqp) [reply libavoidrepetition.h:171]: Do we go through add these `strcpy` hassle and  dynamic `realloc` hassle every time we wanna add 2 strings?

No. We don't. So, let's hide that with an abstraction `add()` that just adds!

Comment (koreaneggroll) [reply libavoidrepetition.h:171]: alright then, ima do it

**Pull request closed without merging**